### PR TITLE
Add MegatronGPTDeployable for serving .nemo models in pytriton

### PR DESCRIPTION
**Pull request by jukim-nv**
# What does this PR do ?

We currently only support deploying TRTLLM models in the Triton inference server.  This MR adds the necessary support to let us deploy .nemo models to Triton as well.

**Collection**: deploy

# Changelog 
- Adds MegatronLLMDeployable class which implements ITritonDeployable

# Usage
* Usage is shown in the new tests/deploy/test_triton_deployable.py

```python
    megatron_deployable = MegatronLLMDeployable(args.nemo_checkpoint, args.num_gpus)
    nm = DeployPyTriton(
        model=megatron_deployable,
        triton_model_name=model_name,
        triton_model_version=1,
        max_batch_size=8,
        port=8000,
        address="0.0.0.0",
        streaming=False,
    )
    nm.deploy()
    nm.run()
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/deploy/tritondeployables/megatrongpt_deployable.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+try:
+    from pytriton.decorators import batch
+    from pytriton.model_config import Tensor
+except Exception:
+    use_pytriton = False
+
+LOGGER = logging.getLogger("NeMo")
+
+
+class MegatronGPTDeployable(ITritonDeployable):
```
Comment (oyilmaz-nvidia) [inline nemo/deploy/tritondeployables/megatrongpt_deployable.py:39]: Let's change this name to PyTritonDeployableLLM and move this file under nemo/deploy/nlp

File: nemo/deploy/tritondeployables/megatrongpt_deployable.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    def __init__(self, nemo_checkpoint_filepath: str, load_model: bool = True):
```
Comment (oyilmaz-nvidia) [inline nemo/deploy/tritondeployables/megatrongpt_deployable.py:40]: Can we also add support for the models in the memory?

File: tests/deploy/test_triton_deployable.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+def get_args():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description=f"Deploy nemo models to Triton and benchmark the models",
+    )
+
+    parser.add_argument(
+        "--model_name", type=str, required=True,
+    )
+    # parser.add_argument(
```
Comment (oyilmaz-nvidia) [inline tests/deploy/test_triton_deployable.py:65]: Let's rename this file as test_pytriton_deploy

**Review by oyilmaz-nvidia — CHANGES_REQUESTED** [review]

**Review by oyilmaz-nvidia — CHANGES_REQUESTED** [review]
Can we also add this pytorch level deployment in this script https://github.com/NVIDIA/NeMo/blob/main/scripts/deploy/nlp/deploy_triton.py please?

File: nemo/deploy/tritondeployables/megatrongpt_deployable.py (hunk 1 above)
Comment (jukim-nv) [reply nemo/deploy/tritondeployables/megatrongpt_deployable.py]: Renamed to `MegatronLLMDeployable` for now.  As discussed offline, whatever name we decide on for this class should match the Model class that we create inside of it.  So if the final name is `MegatronLLMDeployable`, the internal self.model should be of type `MegatronLLMModel`.

File: nemo/deploy/tritondeployables/megatrongpt_deployable.py (hunk 2 above)
Comment (jukim-nv) [reply nemo/deploy/tritondeployables/megatrongpt_deployable.py]: added the existing_model field to the `MegatronLLMDeployable` constructor

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
+from nemo.collections.nlp.modules.common.text_generation_utils import (
+    OutputType,
+    get_default_length_params,
+    get_default_sampling_params,
+)
+from nemo.collections.nlp.modules.common.transformer.text_generation import LengthParam, SamplingParam
+from nemo.collections.nlp.parts.nlp_overrides import NLPDDPStrategy, NLPSaveRestoreConnector
```
[the remark this answers is not in the document]
Comment (oyilmaz-nvidia) [reply nemo/deploy/nlp/megatronllm_deployable.py:31]: Please resolve all the CodeQL reviews

File: scripts/deploy/nlp/deploy_triton.py [hunk 5]
```diff
@@ -128,18 +145,7 @@
...
-    if args.debug_mode:
-        loglevel = logging.DEBUG
-    else:
-        loglevel = logging.INFO
-
-    LOGGER.setLevel(loglevel)
-    LOGGER.info("Logging level set to {}".format(loglevel))
-    LOGGER.info(args)
-
+def get_trtllm_deployable(args):
```
[the remark this answers is not in the document]
Comment (oyilmaz-nvidia) [reply scripts/deploy/nlp/deploy_triton.py:148]: This one too

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+class MegatronLLMDeployable(ITritonDeployable):
+    """Triton inference server compatible deploy class for a .nemo model file"""
+
+    def __init__(
+        self,
+        nemo_checkpoint_filepath: str = None,
+        num_devices: int = 1,
+        num_nodes: int = 1,
+        existing_model: MegatronGPTModel = None,
```
[the remark this answers is not in the document]
Comment (jukim-nv) [reply nemo/deploy/nlp/megatronllm_deployable.py:99]: Is there a particular convention we are trying to satisfy with renaming to `model`?  I personally don't think that `model` alone is descriptive enough for the argument, considering that the word "model" can refer to many different things the context of AI and LLMs.
`existing_model` is not particularly good either so I would be open to other suggestions.  Maybe `preinitialized_model`?

[the remark this answers is not in the document]
Comment (oyilmaz-nvidia) [reply nemo/deploy/nlp/megatronllm_deployable.py:99]: We basically pass a model in the memory. That's why I was suggesting the model. But I also agree that might be descriptive enough. maybe initialized_model?

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+            trainer = Trainer(
+                strategy=NLPDDPStrategy(),
+                devices=num_devices,
+                num_nodes=num_nodes,
+            )
+
+            custom_config = MegatronGPTModel.restore_from(
+                nemo_checkpoint_filepath, trainer=trainer, return_config=True
+            )
+            # transformer_engine should always be true according to EricH, but GPT-2B model will fail if it is enabled
```
Comment (oyilmaz-nvidia) [inline nemo/deploy/nlp/megatronllm_deployable.py:131]: Let's remove the `according to EricH, but GPT-2B model will fail if it is enabled` part and assume that it should be set to true for now.

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+            custom_config.transformer_engine = True
+            # using multi-gpu for tensor parallelism directly for now, could do pipeline parallel instead or a combination
```
Comment (oyilmaz-nvidia) [inline nemo/deploy/nlp/megatronllm_deployable.py:133]: let's use nemo logger.info here

**Review by oyilmaz-nvidia — DISMISSED** [review]

**Pull request closed without merging**